source/dofs/dof_handler_policy.cc

```diff
@@ -289,44 +429,706 @@ namespace internal
+                                        ExcInternalError());
+
+                                dof_identities[slave_dof_index] = master_dof_index;
+                              }
+                          }
+                    }
+                }
+
+          // finally restore the user flags
+          const_cast<dealii::Triangulation<dim,spacedim> &>(dof_handler.get_triangulation())
+          .load_user_flags_quad(user_flags);
+
+          return dof_identities;
+        }
+
+
+
+        /**
+         * Once degrees of freedom have been distributed on all cells, see if
+         * we can identify DoFs on neighboring cells. This function does nothing
+         * on regular DoFHandlers, but goes through vertices, lines, and quads
+         * for hp::DoFHandler objects.
+         *
+         * Return the final number of degrees of freedom, which is the old one
+         * minus however many were identified
+         */
+        template <int dim, int spacedim>
+        static
+        unsigned int
+        unify_dof_indices (const DoFHandler<dim,spacedim> &,
+                           const unsigned int              n_dofs_before_identification)
+        {
+          return n_dofs_before_identification;
+        }
+
+
+
+        template <int dim, int spacedim>
+        static
+        unsigned int
+        unify_dof_indices (hp::DoFHandler<dim,spacedim> &dof_handler,
+                           const unsigned int            n_dofs_before_identification)
+        {
+          // compute the constraints that correspond to unifying
+          // dof indices on vertices, lines, and quads. do so
+          // in parallel
+          std::map<types::global_dof_index, types::global_dof_index>
+          constrained_vertex_indices, constrained_line_indices, constrained_quad_indices;
+
+          {
+            Threads::TaskGroup<> tasks;
+
+            tasks += Threads::new_task ([&]()
+            {
+              constrained_vertex_indices = compute_vertex_dof_identities (dof_handler);
+            });
+            tasks += Threads::new_task ([&]()
+            {
+              constrained_line_indices = compute_line_dof_identities (dof_handler);
+            });
+            tasks += Threads::new_task ([&]()
+            {
+              constrained_quad_indices = compute_quad_dof_identities (dof_handler);
+            });
+
+            tasks.join_all ();
+          }
+
+          // create a vector that contains the new DoF indices; first preset the
+          // ones that are identities as determined above, then enumerate the rest
+          std::vector<types::global_dof_index>
+          new_dof_indices (n_dofs_before_identification, numbers::invalid_dof_index);
+
+          for (const auto &constrained_dof_indices :
+          {
+            &constrained_vertex_indices,
+            &constrained_line_indices,
+            &constrained_quad_indices
+          })
+          for (const auto &p : *constrained_dof_indices)
```

Comment:
Would you please surround this for loop with extra brackets? I found it really confusing that this starts on the same line as the previous for loop.